### ISSUE TITLE
Wait on multiple condition variables?

### ISSUE THREAD

**Issue by wlandau** [shikokuchuo/nanonext#17]
I really like `nextget("cv")` in `mirai`, and I think it could make polling in `targets` more gentle. However, in `crew`, there is a feature called [controller groups](https://wlandau.github.io/crew/articles/groups.html) which aggregates multiple controllers under a single interface. This feature meets the request from https://github.com/shikokuchuo/mirai/discussions/28 for heterogeneous workers. Each controller in the controller group has its own condition variable, and it would be great to be able to wait until at least one of these condition variables gets signaled.

**Comment by shikokuchuo** [comment]
Excellent - you already know that this feature request belongs at this level!

However, I wonder if 'brute force' implementation is the correct answer here...

If in a controller group, would all waits move to the group level? If so then you can already hack the behaviour by assigning the cv in the compute environments to be the same one (it's an external pointer - reference object). For testing at least. I don't suggest you do this for the release yet, until there's a proper interface at least!

**Comment by wlandau** [comment]
> However, I wonder if 'brute force' implementation is the correct answer here...

I wish there were a different way, but I am not sure how. The only alternative I can think of is to wait on each condition variable sequentially, which would not be efficient. If a task resolves in controller B but not controller A, then I might incorrectly spend the full timeout interval waiting for A.

**Comment by wlandau** [comment]
> If in a controller group, would all waits move to the group level? If so then you can already hack the behaviour by assigning the cv in the compute environments to be the same one (it's an external pointer - reference object). For testing at least. I don't suggest you do this for the release yet, until there's a proper interface at least!

Yes, it's at the group level. But the implementation will be so easy on my end that I can wait for this #17 before I get started.

Once I can use condition variables, both `crew` and `targets` will be a whole lot faster and much kinder to the CPU. Awesome things to look forward to if feasible in `nanonext`.

**Comment by shikokuchuo** [comment]
I'm a bit confused by your last 2 replies. If it works to have the entire group share the same cv, then this can be implemented in mirai as a function that assigns the cv of one compute profile to another or something like that. This would not require a new function at nanonext.

**Comment by wlandau** [comment]
I think I see what you mean: just have all tasks signal to the same cv upon resolution. But I am not sure that would be compatible with https://github.com/shikokuchuo/mirai/discussions/78#discussioncomment-7148551. `crew` now uses `nextget("cv")` to count the resolved tasks of each individual controller separately, and that mechanism is critical to auto-scaling at the level of each controller.

**Comment by shikokuchuo** [comment]
First, condition variables are meant for synchronisation - that is you wait for one to be signalled and then something happens. Now just because there is a counter in there, using it to count is just an incidental by-product - it's not a feature so to speak. So I'm not surprised if using it to synchronise vs. to count will be incompatible - this is not a deficiency.

As for waiting for 2 at the same time, that is not implemented by definition of what these are. One solution could be to have 3 condition variables and 2 threads. You wait on one of them (group controller level), and then have 2 threads waiting on each of the others (one for each controller). The side-effect of the cvs being signalled is then for them to signal in turn the one you are waiting on.

**Comment by shikokuchuo** [comment]
As a 'conditionVariable' is an external pointer, we may be able to use an infix operator like:
```r
cv %<~% cv2
```
Which would mean the signal from `cv2` is forwarded to `cv`. So every time `cv2` is signalled, so is `cv`.

You could forward all the cvs within the controller group to  one of them and just wait on the one condition variable. Does this achieve what you want?

**Comment by wlandau** [comment]
> Now just because there is a counter in there, using it to count is just an incidental by-product - it's not a feature so to speak. 

Do you think there is a risk that CVs would drop their nice counting properties at some point?

> You could forward all the cvs within the controller group to one of them and just wait on the one condition variable. Does this achieve what you want?

Yes, that would be perfect! Is it already possible?

**Comment by shikokuchuo** [comment]
> Do you think there is a risk that CVs would drop their nice counting properties at some point?

No, it's inherent to what they are.

> 
> > You could forward all the cvs within the controller group to one of them and just wait on the one condition variable. Does this achieve what you want?
> 
> Yes, that would be perfect! Is it already possible?

It's just a thought experiment. It's a pretty naive way to implement this. Possibly I'll be putting in place something else that also achieves what you want.
